### DISCUSSION TITLE
Advanced custom shape that can change based on preset options?

### DISCUSSION THREAD

**Discussion by rezanid** [jgraph/drawio#2466] [Ideas]
I was thinking about building a set of custom shapes that can change / switch their look when:
* users drag yellow handles of the shape or 
* users select preset options (e.g. using custom data) or
* users connect certain kind of shapes together

I know that MS Visio shapes for example provides these kind of features through its Excel like sheets (not so user-friendly though), but I was wondering if the same is achievable in Draw IO.

I already read through your blog posts, but was not able to find a tutorial that goes through A-Z of a real advanced shape like some of the out-of-the-box ones you have in the app.

**Comment by davidjgraph** [comment]
There's no developer documentation. You're only route is to take a shape that does something similar and adapt it.

**Comment by Peeeet** [comment]
I am with rezanid on this one, I would love to be able to make my own advanced shapes, and while I'm not afraid to dig deep and look under the hood to try to do this by adapting another shape or even using those as starting points to figure out how to do that, I can't seem to even figure out how to use a custom shape I would have created.

For example, as far as I can tell, the way that all the built-in shapes are "defined" is in js and xml files from the "shapes" and "stencils" folders in src/main/webapp, though inside those files, I see mentions of extending mxgraph (which if referring to jgraph/mxgraph repo is EOL right?) and also "prototypes" which I have no clue where those might be (if that's even a thing)  And then I do see a couple files in src/main/webapp/js that might be pertinent, but they are minified so I assume are just for runtime (I am NOT a developer if you haven't noticed)

All I would ask for is maybe a hint or point in the right direction of WHICH files we should be looking at, and even if not that (as I assume even mildly experienced JS devs could probably tell me that) maybe just a hint as to, if one were to be able to make custom libraries similar to the native ones, how / where would one even put those files in a private / self-hosted install (or even the desktop app - though I assume that would require compiling something) to make them accessible to the program.

I'm very happy for the software as it is and I love that it is free / open source (I would pay for it) but I feel there is so much more that semi-advanced users like rezanid could accomplish if there was some level between just using the software and full-on forking it to try to re-invent the wheel.  I don't even think it needs to be a full tutorial or even any sort of official / supported documentation, just a hint for people who have been digging as to where to even start to start looking, and a "on your own" warning after that.

I probably couldn't be that person myself, but I have found that it only takes one new / fresh person to do some of the legwork and then the rest of us intermediaries can usually stumble our way through and get some very useful results out of it - similar to how many gaming modders can often make better games / features than the original with even a crumb of behind-the-scenes information.
